alternator, test: add reproducer for issue about immediate LWT timeout

nyh opened the pull request:
This patch adds a reproducer for issue #16261, where it was reported that when Alternator read-modify-write (using LWT) operations to the same partition are sent to different nodes, sometimes the operation fails immediately, with an InternalServerError claiming to be a "timeout", although this happens almost immediately (after a few milliseconds), not after any real timeout.

The test uses 3 nodes (so it is in the topology framework), and 3 threads which send RMW operations to different items in the same partition, and with almost certainty reaches the InternalServerError in around 100 writes by each thread. This InternalServerError looks like:

    Internal server error: exceptions::mutation_write_timeout_exception
    (Operation timed out for alternator_alternator_Test_1719157066704.alternator_Test_1719157066704 - received only 1 responses from 2 CL=LOCAL_SERIAL.)

The test also prints how much time it took for the request to fail, for example:
    In incrementing 1,0 on node 1: error after 0.017074108123779297
This is 0.017 seconds - it's not the cas_contention_timeout_in_ms timeout (1 second) or any other timeout.

If we enable trace logging, adding to topology_experimental_raft/suite.yaml
    extra_scylla_cmdline_options: ["--logger-log-level", "paxos=trace"]
we get the following TRACE-level message in the log:

    paxos - CAS[0] accept_proposal: proposal is partially rejected

This again shows the problem is "uncertainty" (partial rejection) and not a timeout.

Refs #16261

**Please replace this line with justification for the backport/\* labels added to this PR**

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9908/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9908/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9908/)
#### Build Details:
- Duration: 2 hr 12 min
- Builder: spider6.cloudius-systems.com

nyh commented:
Rebased to fix merge conflicts. Please review.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12799/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12799/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12799/)
#### Build Details:
- Duration: 5 hr 57 min
- Builder: spider2.cloudius-systems.com

avikivity commented on test/topology_experimental_raft/test_alternator.py:
```diff
@@ -522,3 +524,83 @@ async def test_alternator_enforce_authorization_true(manager: ManagerClient):
+    def run(self):
+        try:
+            self.ret = self._target(*self._args, **self._kwargs)
+        except BaseException as e:
```
Wouldn't asyncio.to_thread be sufficient?

nyh replied:
I don't know. As I noted, all the tests in test_alternator.py use boto3 doesn't which doesn't support asyncio at all. So all the tests just **ignore** asyncio and use normal threaded code. So unsurprisingly, when I need to run _another_ thread, I use the normal Python threading.Thread to create one - not anything coming from asyncio.

The reason why I needed to use that ThreadWrapper around threading.Thread instead of using the latter directly is explained in the comment - threading.Thread doesn't do the right thing when there is an error inside the thread (we need the error to fail the entire test, not just the single thread it happened in).

nyh commented:
Rebased to fix conflicts and run the CI again with a refreshed master.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14563/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/14563/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
#### Build Details:
- Duration: 6 hr 5 min
- Builder: spider4.cloudius-systems.com

nyh commented:
Rebased. The directory structure changed considerably, and the "alternator3" fixture is gone (after it didn't work correctly) so fix the test create its own 3-node cluster. Changed the commit message accordingly (I updated the commit message of the single commit, didn't bother to change the cover letter here as it will be ignored anyway).

I verified that without the xfail marker, the test still (usually, but not always) fails as expected (unfortunately, test.py doesn't have pytest's "--runxfail" option).

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15900/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/15900/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: alternator
#### Build Details:
- Duration: 3 hr 26 min
- Builder: i-0faf03494b6f9c272 (m5ad.12xlarge)

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

| Mode | Stage | Status | Node |
|------|-------|--------|------|
|  | [Framework test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18083/) | :white_check_mark: | i-0f54e4e84f50d15e8 |
| dev | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18083/) | :white_check_mark: | i-06bf0dd9f836bbb44 |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18083/) | :white_check_mark: | i-06bf0dd9f836bbb44 |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18083/)<br> :small_blue_diamond: alternator | :white_check_mark: | i-06bf0dd9f836bbb44 |
| release | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18083/) | :white_check_mark: | i-01184a07c41416bc0 |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18083/) | :white_check_mark: | i-01184a07c41416bc0 |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18083/)<br> :small_blue_diamond: alternator | :white_check_mark: | i-01184a07c41416bc0 |
| debug | [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18083/) | :white_check_mark: | i-0d37cd62602647df1 |
|  | [test.py](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18083/) | :white_check_mark: | i-0d37cd62602647df1 |
|  | [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18083/)<br> :small_blue_diamond: alternator | :white_check_mark: | i-0d37cd62602647df1 |

#### Build Details:
- Duration: 20 hr

denesb commented:
Wow, a PR old enough to have its own anniversary.
Queued.
